### PR TITLE
Kubectl taint node based on label selector

### LINKED ISSUE #44522: Label selector not supported in kubectl taint
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): taint, selector

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:36:33Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1", GitCommit:"b0b7a323cc5a4a2019b2e9520c21c7830b7f708e", GitTreeState:"clean", BuildDate:"2017-04-03T20:33:27Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**: GKE
- **OS** (e.g. from /etc/os-release): MacOS client, linux server
- **Kernel** (e.g. `uname -a`):
- **Install tools**:gcloud
- **Others**:

**What happened**:
I tried to kubectl taint some nodes using a label selector:
kubectl taint nodes -l dedicated=foo dedicated=foo:PreferNoSchedule

and got the error:
error: at least one resource name must be specified since 'all' parameter is not set

**What you expected to happen**:
Those nodes would be tainted with dedicated=foo:PreferNoSchedule.  Trying to use --all along with the selector is also not supported.

**How to reproduce it** (as minimally and precisely as possible):
kubectl taint nodes -l <any label selector> dedicated=foo:PreferNoSchedule

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes  #44522
**Release note**:
```
Taints the node based on label selector
```

### CODE REVIEW

**Comment by jayunit100** [comment]
Thanks @ravisantoshgudimetla this works for me :) 

```
cluster/kubectl.sh label nodes --all "myLabel=x" ; cluster/kubectl.sh taint nodes --selector='myLabel==x' dedicated=foo:PreferNoSchedule ;
node "127.0.0.1" labeled
node "127.0.0.1" tainted
```

Do we need the unit tests in order to merge this ? I'd prefer to push it forward since the unit tests (we think) don't seem to support combining kubectl labelling with tainting (although maybe there is way to do so, its just not obvious).

**Comment by ravisantoshgudimetla** [comment]
/cc @jayunit100. 

- What is lacking?
Writing a uni test seems problematic, as the fake client is not able to build the information from labels.  

- What should we do?
So, the ideal way would be to have arguments to RunTaint(factory etc), fill in the values in test suite and fill in TaintOptions{} struct and build the information in test case. As this whole file needs some re-ordering(as discussed in #43171) the test-cases should be based on https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/taint.go#L362, if the node is tainted or not. The current approach seems to be checking from API object of node.(which should be in other file like TaintController etc).

**Comment by ravisantoshgudimetla** [comment]
cc @aveshagarwal

**Comment by jayunit100** [comment]
@k8s-bot test this please

**Comment by jayunit100** [comment]
cc @nimeshksingh

File: pkg/kubectl/cmd/taint.go [hunk 1]
```diff
@@ -248,12 +254,12 @@ func (o *TaintOptions) Complete(f cmdutil.Factory, out io.Writer, cmd *cobra.Com
 	o.builder = resource.NewBuilder(mapper, f.CategoryExpander(), typer, resource.ClientMapperFunc(f.ClientForMapping), f.Decoder(true)).
 		ContinueOnError().
 		NamespaceParam(namespace).DefaultNamespace()
+	if o.selector != "" {
+		o.builder = o.builder.SelectorParam(o.selector).ResourceTypes("node")
+	}
 	if o.all {
 		o.builder = o.builder.SelectAllParam(o.all).ResourceTypes("node")
 	} else {
-		if len(o.resources) < 2 {
```
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: @aveshagarwal ok to get rid of this check?

File: pkg/kubectl/cmd/taint.go [hunk 2]
```diff
@@ -80,7 +80,13 @@ var (
 		kubectl taint nodes foo dedicated:NoSchedule-
 
 		# Remove from node 'foo' all the taints with key 'dedicated'
-		kubectl taint nodes foo dedicated-`))
+		kubectl taint nodes foo dedicated-
+
+		# Add from node 'foo' all the taints with key 'dedicated'
+		kubectl taint nodes foo dedicated-
```
Comment (supereagle) [inline pkg/kubectl/cmd/taint.go]: The commands for `Add` and `Remove` examples are the same. Any error?

File: pkg/kubectl/cmd/taint.go (hunk 1 above)
Comment (aveshagarwal) [inline pkg/kubectl/cmd/taint.go]: It seems No. if both "selector" and "all"  are NOT provided, we should do this check to verify the sequence of <resource> <name>.

File: pkg/kubectl/cmd/taint.go (hunk 2 above)
Comment (aveshagarwal) [inline pkg/kubectl/cmd/taint.go]: It seems incorrect, the format with hyphen in the end ("dedicated-") is only for deletion. It does not make sense with addition.

Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: Right. Its a mistake, Thanks for catching it @supereagle & @aveshagarwal , I will change this.

**Comment by kargakis** [comment]
@kubernetes/sig-cli-pr-reviews ptal

File: pkg/kubectl/cmd/taint.go (hunk 1 above)
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: @aveshagarwal https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/taint.go#L236 isn't sufficient ?

Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: @aveshagarwal  So, in the below command, I have a resource `node` without selector and all options. Its throwing a valid error. 
`cluster/kubectl.sh taint node 127.0.0.1 `
`error: at least one taint update is required`

Comment (aveshagarwal) [inline pkg/kubectl/cmd/taint.go]: I meant this errneous case:
kubectl  taint node key:value:Noschedule

Comment (aveshagarwal) [inline pkg/kubectl/cmd/taint.go]: IOW, if you dont specify node name, but just the resource type node.

Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: Ok, thanks for the information @aveshagarwal. So, we already have a check at API server to handle it. There is no need to validate it here again.

cluster/kubectl.sh taint node dedicated=foo:PreferNoSchedule
error: You must provide one or more resources by argument or filename.
Example resource specifications include:
   '-f rsrc.yaml'
   '--filename=rsrc.json'
   'pods my-pod'
   'services'

Comment (aveshagarwal) [inline pkg/kubectl/cmd/taint.go]: This is part of command line validation, and should be done in kubectl too.

Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: So, if that is the case, it should be something like if(o.selector)!=="" and o.all=false then we should throw this error. I can make that change.

Comment (aveshagarwal) [inline pkg/kubectl/cmd/taint.go]: Sound good. And as you are at this, make sure one of `selector` and `all`  is present not both simultaneously if not already.

**Comment by ravisantoshgudimetla** [comment]
@aveshagarwal PTAL.

**Comment by aveshagarwal** [comment]
@ravisantoshgudimetla lgtm, now need to figure out the issue with your unit test.

**Comment by jayunit100** [comment]
@kubernetes/sig-cli-pr-reviews  Can we merge this, it fixes a pretty important label + kubectl bug  ? @ravisantoshgudimetla can you create a follow on issue for adding a unit test.

**Comment by derekwaynecarr** [comment]
the code looks fine to me, for a unit test, we appear to have testing for label selectors in delete_test.go that may have a convention you could follow in `TestDeleteMultipleSelector`.  if there is no urgency, would prefer a test case with the code in one pr.

**Comment by ravisantoshgudimetla** [comment]
@derekwaynecarr Thanks for the pointer regarding test case. I will check that and get back in case I am stuck.

**Comment by ravisantoshgudimetla** [comment]
@derekwaynecarr I tried running the cmd using the way in deleted_test.go and it is causing the same error. The test case that is failing is https://gist.github.com/ravisantoshgudimetla/d66fb36b1279f118395e540cc8c61505#file-taint_test-go-L106.

As updated earlier, It seems the issue is with object being built using builder. The ones coming from unit tests are not having name created at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/taint.go#L320. So, wondering if there is a way to work around that to write a unit  test. Initially I was thinking of mopdifying RunTaint() to take some arguments(especially for builders). Please lmk, if you have other thoughts or if I missed anything.

**Comment by jayunit100** [comment]
@derekwaynecarr  @ravisantoshgudimetla ok so sounds like to move forward, we have to do two things... 

1.  Lets create a follow on issue delineating the lack of the builder mocks needed for better higher level tests, and aim  to improve the general situation for `kubectl` tests in general, so that in the future we can more declaratively test the subtleties of taints w/ different builder+label+... situations and so on.  
2. per @derekwaynecarr suggestion , we can make a unit test for this patch by making a separate function for object creation.  

(derek let us know if this sounds good to you as well, i think that will solve everything brought up in this issue, including getting this much needed fix merged :))

**Comment by derekwaynecarr** [comment]
i have no strong opinion on the style of the test, just would like a test of some kind.  i am not sure where we are these days with integration tests via kubectl but if that is easier than unit testing that is fine too i suspect.

**Comment by ravisantoshgudimetla** [comment]
@derekwaynecarr @jayunit100, 
- I separated out the logic for flag validation and added a test case that validates it but for object builder mocking, I raised another issue.
- I have modified the existing test case to have a field called selector, so that it would be easier for extending later.

Please lmk, if you have any other concerns.

File: pkg/kubectl/cmd/taint_test.go [hunk 3]
```diff
@@ -332,3 +334,50 @@ func TestTaint(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateFlags(t *testing.T) {
```
Comment (jayunit100) [inline pkg/kubectl/cmd/taint_test.go]: yeah this is way cleaner, thanks ravi

File: pkg/kubectl/cmd/taint_test.go [hunk 4]
```diff
@@ -332,3 +334,50 @@ func TestTaint(t *testing.T) {
...
+			expectFatal: false,
+		},
+		{
+			taintOpts:   TaintOptions{selector: "myLabel=X", all: true},
+			description: "With selector and with all flags",
+			expectFatal: true,
+		},
+		{
+			taintOpts:   TaintOptions{selector: "", all: false, resources: []string{"node"}},
+			description: "With both selector and all flags and if node name is not provided",
```
Comment (jayunit100) [inline pkg/kubectl/cmd/taint_test.go]: fix desc

**Comment by jayunit100** [comment]
I think the test is sufficient to prevent regressions now... so i guess lets move this forward and follow up on #44953 to mock these out properly for better kubectl tests........  

PTAL @derekwaynecarr @aveshagarwal

**Comment by aveshagarwal** [comment]
lgtm

**Comment by ravisantoshgudimetla** [comment]
@jayunit100 - PTAL, I have addressed your comments on descriptions.

**Comment by ravisantoshgudimetla** [comment]
@jayunit100 I did a gofmt again. Can you please ask the bot test?

**Comment by ravisantoshgudimetla** [comment]
and the unit tests seem to be a flake..

**Comment by jayunit100** [comment]
@k8s-bot ok to test

**Comment by ravisantoshgudimetla** [comment]
#45170  - Flake..

**Comment by jayunit100** [comment]
@k8s-bot unit test this please

**Comment by jayunit100** [comment]
@kubernetes/sig-cli-maintainers can we get an approve on this ? Thanks :)

**Pull request merged**